// @vlcn.io/ws-client/dist/transport/WebSocketTransport.js
import { decode, encode, tags, } from "@vlcn.io/ws-common";
export default class WebSocketTransport {
    #socket = null;
    #hadStartStream = false;
    #closed = false;
    #options;
    #onReady = null;
    #keepAliveInterval = null;
    constructor(options) {
        this.#options = options;
    }
    start(onReady) {
        this.#onReady = onReady;
        this.#socket = this.#openSocketAndKeepAlive(this.#options);
    }
    #openSocketAndKeepAlive(options) {
        if (this.#closed) {
            return null;
        }
        if (this.#keepAliveInterval == null) {
            this.#keepAliveInterval = setInterval(() => {
                if (!this.#socket || this.#socket?.readyState === WebSocket.CLOSED) {
                    this.#openSocketAndKeepAlive(options);
                }
            }, Math.random() * 2000 + 1000);
        }
        const socket = new WebSocket(options.url, [
            btoa(`${options.authToken != null ? `auth=${options.authToken},` : ""}room=${options.room}`).replaceAll("=", ""),
        ]);
        socket.binaryType = "arraybuffer";
        socket.addEventListener("message", (e) => {
            this.#processEvent(new Uint8Array(e.data));
        });
        socket.onopen = () => {
            if (this.#onReady)
                this.#onReady();
        };
        this.#socket = socket;
        return socket;
    }
    onChangesReceived = null;
    onStartStreaming = null;
    onResetStream = null;
    onReconnected = null;
    #processEvent = (data) => {
        const msg = decode(data);
        switch (msg._tag) {
            case tags.AnnouncePresence:
            case tags.RejectChanges:
                // clients should not receive these evnts
                throw new Error(`Unexpected event: ${msg._tag}`);
            case tags.Changes:
                if (this.onChangesReceived) {
                    this.onChangesReceived(msg);
                }
                return;
            case tags.StartStreaming:
                if (this.#hadStartStream) {
                    this.onResetStream && this.onResetStream(msg);
                }
                else {
                    this.#hadStartStream = true;
                    this.onStartStreaming && this.onStartStreaming(msg);
                }
                return;
        }
    };
    announcePresence(msg) {
        this.#socket.send(encode(msg));
    }
    sendChanges(msg) {
        if (this.#socket.readyState !== WebSocket.OPEN) {
            return "reconnecting";
        }
        // If we do the below we need to nofiy the caller to back off on sending.
        if (this.#socket.bufferedAmount > 1024 * 1024 * 5) {
            console.warn("socket buffer full. Waiting till buffer is drained before allowing more changes to be queue for send.");
            return "buffer-full";
        }
        this.#socket.send(encode(msg));
        return "sent";
    }
    rejectChanges(msg) {
        this.#socket.send(encode(msg));
    }
    close() {
        this.#closed = true;
        this.#socket.close();
        if (this.#keepAliveInterval) {
            clearInterval(this.#keepAliveInterval);
        }
    }
}
